### PR TITLE
ACL: Fix Multi Module Tab Inclusion

### PR DESCRIPTION
**Pull request by ryanrath**
## Description
- This commit should resolve the issue where modules add information to the
  'default' role profile and those updates are not picked up for inclusion into
  the database.

## Motivation and Context
Tabs are good, modules should be able to add tabs.

## Tests performed
Manual Testing was performed: 

- Spin up a docker container from: tas-tools-ext-01.ccr.xdmod.org/centos7-xdmod-supremm6.6.0:version3
- `export XDMOD_TEST_MODE=fresh_install`
- `git clone --branch xdmod7.1 https://github.com/ubccr/xdmod.git ~/src/github.com/ubccr/xdmod`
- `git clone --branch xdmod7.1 https://github.com/ubccr/xdmod-appkernels.git ~/src/github.com/ubccr/xdmod-appkernels`
  - **NOTE: I had to update the xdmod-appkernels.spec.in file so that it would allow xdmod versions <= xdmod7.1 ( previously it was < xdmod7.1 )**
- `git clone --branch xdmod7.1 https://github.com/ubccr/xdmod-supremm.git ~/src/github.com/ubccr/xdmod-supremm`
- `git clone --branch acl/fix_multi_module_tabs https://github.com/ryanrath/xdmod.git ~/src/github.com/ryanrath/xdmod`
- `cd ~/src/github.com/ubccr/xdmod-appkernels`
- `ln -s ``pwd`` ../xdmod/open_xdmod/modules/appkernels`
  - NOTE: I couldn't for the life of me get a single back tick to work inside of a code block. 
- `ln -s ``pwd`` ~/src/github.com/ryanrath/xdmod/open_xdmod/modules/appkernels`
- `cd ../xdmod-supremm`
- `ln -s ``pwd`` ../xdmod/open_xdmod/modules/supremm`
- `ln -s  ``pwd`` ~/src/github.com/ryanrath/xdmod/open_xdmod/moduels/supremm`
- `composer install -d ../xdmod`
- `cd ../xdmod`
- `~/bin/buildrpm xdmod appkernels supremm`
- `cd ../xdmod-supremm`
- `./tests/integration_tests/scripts/bootsrap.sh`
- `xdmod-setup`
  - Option 2.) Database
  - Fill in the DB information
  - Wait for the acl scripts to finish
- open up a browser and direct it to the newly installed XDMoD
- log in as a user ( doesn't matter which one )
- The App Kernels & SUPREMM tab should be missing
- go back to the docker container
- `cd ~/src/github.com/ryanrath/xdmod`
  - **NOTE: I  bumped the release version in open_xdmod/modules/xdmod/build.json so that I could yum install the newly built rpm**
- `~/bin/buildrpm xdmod`
- `yum -y localinstall ~/rpmbuild/RPMS/noarch/xdmod-7.1.0-1.1.el7.centos.noarch.rpm`
- `xdmod-setup` ( like last time )
  - Option 2.) Databse
  - Fill in the DB information
  - Wait for the acl scripts to finish
- open up a browser and navigate to the newly upgraded XDMoD
- log in as a user ( doesn't matter which one )
- The App Kernels and SUPREMM tabs should now be present and functional 

Manual Database Tests:
-  start the mysql client inside the docker container
- `use moddb;`
- `SELECT * FROM modules;`
  - This should return 3 entries: appkernels, supremm, and xdmod
- `SELECT * FROM realms;`
  - This should return 2 entries: Jobs, SUPREMM
- `SELECT * FROM acls;`
  - This should return 6 entries: pub, usr, cd, pi, cs, mgr
- `SELECT * FROM tabs;`
  - This should return 7 entries: tg_summary, tg_usage, about_xdmod, metric_explorer, report_generator, app_kernels, and job_viewer.
- `SELECT a.name, t.name FROM acl_tabs at JOIN acls a ON a.acl_id = at.acl_id JOIN tabs t ON t.tab_id = at.tab_id;`
  - This should return 38 records and should correspond to the `permitted_modules` entries in roles.json
  - The important thing to note is that `pub` does not have access to app_kernels / job_viewer and that the other acls do. 
- If you want to check group_bys or statistics ( information in datawarehouse.json ) you can run the following ( but really these records should not be affected by the change ): 
  - `SELECT m.name AS module, r.name AS realm, gb.name AS group_by FROM group_bys gb LEFT JOIN modules m ON m.module_id = gb.module_id LEFT JOIN realms r ON r.realm_id = gb.realm_id;`
  - `SELECT m.name AS module, r.name AS realm, s.name AS statistic FROM statistics s LEFT JOIN modules m ON m.module_id = s.module_id LEFT JOIN realms r ON r.realm_id = s.realm_id;`

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
)

### CODE REVIEW

File: bin/acl-config [hunk 1]
```diff
@@ -914,6 +944,16 @@ SQL;
             $log->info("[FAIL] Inserted more than one Acl Hierarchy Record for: $id");
         }
     }
+    $committed = $db->commit();
+    if ($committed === false) {
+        $log->warning("[FAIL] Attempt to commit hierarchies for Acl [$acl] failed.");
+        $rolledBack = $db->rollBack();
+
+        // We throw an Exception so that we do not leave the db in an inconsistent state.
```
Comment (jpwhite4) [inline bin/acl-config]: This comment is at odds with the message in the exception. Will the database be inconsistent or not?

**Comment by jpwhite4** [comment]
Did you also test the upgrade mode?

**Comment by jpwhite4** [comment]
The shippable tests will not trigger the bug that you are trying to fix since they do not install the modules. Therefore you cannot use the shippable test results to check whether this bug has been fixed.

**Comment by ryanrath** [comment]
Ok, would the following be acceptable? I'll perform a manual upgrade procedure akin to the test I describe above w/ the change that instead of running `xdmod-setup` after installing the updated xdmod rpm I will instead run `xdmod-upgrade`?

File: bin/acl-config [hunk 2]
```diff
@@ -148,17 +156,27 @@ function main()
...
+        } elseif ($hasModuleRealms){
+            $defaultRealms = $datawarehouseSource[DEFAULT_MODULE_NAME]['realms'];
+            $moduleRealms = $datawarehouseSource[$module]['realms'];
+            $moduleRealmData = array_merge_recursive($defaultRealms, $moduleRealms);
+        } else {
+            continue;
+        }
+
+        foreach ($moduleRealmData as $realmDisplay => $dwData) {
+            $groupBys = null !== $dwData['group_bys'] ? $dwData['group_bys'] : null;
```
Comment (jpwhite4) [inline bin/acl-config]: This code could be simplified to:
```
$groupBys = $dwData['group_bys']
$statistics = $dwData['statistics']
```
Is this what you intented to happen? or is this a bug?

Comment (ryanrath) [reply bin/acl-config]: What I really intended was: 
```php
$groupBys = isset($dwData['group_bys']) ? $dwData['group_bys'] : array();
$statistics = isset($dwData['statistics']) ? $dwData['statistics'] : array();
```
This ensures that we don't attempt to request data from an index that doesn't exist.

File: bin/acl-config [hunk 3]
```diff
@@ -1316,7 +1352,23 @@ SQL;
...
+SQL;
+        $updated = $db->execute(
+            $updateQuery,
+            array(
+                ':realm_display' => $realmDisplay,
+                ':realm_name'    => $realmName
+            )
+        );
+        if ($updated === 1) {
+            $log->info("Updated Realm [$realmName] with display [$realmDisplay]");
```
Comment (smgallo) [inline bin/acl-config]: Maybe include the module name in this one as you did below so they're consistent?

File: bin/acl-config [hunk 4]
```diff
@@ -914,6 +944,16 @@ SQL;
...
         }
     }
+    $committed = $db->commit();
+    if ($committed === false) {
+        $log->warning("[FAIL] Attempt to commit hierarchies for Acl [$acl] failed.");
+        $rolledBack = $db->rollBack();
+
+        // We throw an Exception to indicate that the database may be in an inconsistent state.
+        if ($rolledBack === false) {
+            throw new \Exception("Unable to rollback Acl [$acl] hierarchy records. Database may be left in an inconsistent state.");
```
Comment (smgallo) [inline bin/acl-config]: You are using `\Exception` here and `Exception` above, please be consistent.

File: bin/acl-config [hunk 5]
```diff
@@ -566,7 +582,7 @@ function processModules(iDatabase $db, array $modules)
         $display = isset($data['display']) ? $data['display'] : $name;
         $enabled = isset($data['enabled']) ? $data['enabled'] : true;
         $version = $data['version'];
-        $installedOn =  isset($data['installed_on'])
+        $installedOn = isset($data['installed_on'])
```
Comment (chakrabortyr) [inline bin/acl-config]: Nitpick: easier to focus on the code changes without the whitespace changes.

File: bin/acl-config (hunk 4 above)
Comment (chakrabortyr) [reply bin/acl-config]: w/o backslash is probably better for PHP 7+.

File: bin/acl-config [hunk 6]
```diff
@@ -810,17 +829,23 @@ SQL;
 
     $inserted = $db->execute($query, $params);
     if ($inserted === 0) {
+        // If we did not insert a record then the assumption is that it already
```
Comment (chakrabortyr) [inline bin/acl-config]: comments within the code block in one block and without in the other.

File: bin/acl-config (hunk 2 above)
Comment (chakrabortyr) [reply bin/acl-config]: Will changing these default values from null -> empty array affect behavior anywhere else in the code?

**Comment by ryanrath** [comment]
I'm closing this PR as, after a discussion with @jpwhite4  and @plessbd a better solution seemed to be re-working the way in which Config constructs / provides data in a module specific way. I'll be adding a new PR that makes these changes.

**Pull request closed without merging**